pyadjoint/checkpointing.py

```diff
@@ -0,0 +1,328 @@
+
+        self.tape._eagerly_checkpoint_outputs = cp_action.write_adj_deps
+
+        if timestep > cp_action.n0:
+            if cp_action.write_ics and timestep == (cp_action.n0 + 1):
+                # Stores the checkpointint data in RAM
+                # This data will be used to restart the forward model
+                # from the step `n0` in the reverse computations.
+                self.tape.timesteps[cp_action.n0].checkpoint()
+
+            if not cp_action.write_adj_deps:
+                # Remove unnecessary variables from previous steps.
+                for var in self.tape.timesteps[timestep - 1].checkpointable_state:
+                    var._checkpoint = None
+                for block in self.tape.timesteps[timestep - 1]:
+                    # Remove unnecessary variables from previous steps.
+                    for output in block.get_outputs():
+                        output._checkpoint = None
+
+        if timestep in cp_action:
+            self.tape.get_blocks().append_step()
+            if cp_action.write_ics:
+                self.tape.latest_checkpoint = cp_action.n0
+            return True
+        else:
+            return False
+
+    @process_taping.register(EndForward)
+    def _(self, cp_action, timestep):
+        if timestep != self.timesteps:
+            raise CheckpointError(
+                "The correct number of forward steps has notbeen taken."
+            )
+        self.mode = Mode.EVALUATED
+        return True
+
+    def recompute(self, functional=None):
+        """Recompute the forward model.
+
+        Parameters
+        ----------
+        functional : BlockVariable
+            The functional to be evaluated.
+        """
+        self.mode = Mode.RECOMPUTE
+
+        with self.tape.progress_bar("Evaluating Functional",
+                                    max=self.timesteps) as bar:
+            # Restore the initial condition to advance the forward model
+            # from the step 0.
+            current_step = self.tape.timesteps[self.forward_schedule[0].n0]
+            current_step.restore_from_checkpoint()
+            for cp_action in self.forward_schedule:
+                self._current_action = cp_action
+                self.process_operation(cp_action, bar, functional=functional)
+
+    def evaluate_adj(self, last_block, markings):
+        """Evaluate the adjoint model.
+
+        Parameters
+        ----------
+        last_block : int
+            The last block to be evaluated.
+        markings : bool
+            If True, then each `BlockVariable` of the current block will have
+            set `marked_in_path` attribute indicating whether their adjoint
+            components are relevant for computing the final target adjoint
+            values.
+        """
+        # Work out other cases when they arise.
+        if last_block != 0:
+            raise NotImplementedError(
+                "Only the first block can be evaluated at present."
+            )
+
```

Comment:
Shouldn't this error message be somewhere else than inside `evalute_adj`, as `last_block` is never used inside this function, except for the error checking?